exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+//
+// Note: In the case of MachinePools the machines are created in order to surface in CAPI what exists in the
+// infrastructure while instead on MachineDeployments, machines are created in CAPI first and then the
+// infrastructure is created accordingly.
+// Note: When supported by the cloud provider implementation of the MachinePool, machines will provide a means to interact
+// with the corresponding infrastructure (e.g. delete a specific machine in case MachineHealthCheck detects it is unhealthy).
+func (r *MachinePoolReconciler) reconcileMachines(ctx context.Context, mp *expv1.MachinePool, infraMachinePool *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	var noKind bool
+	var noSelector bool
+
+	var infraMachineKind string
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noKind = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	var infraMachineSelector metav1.LabelSelector
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineSelector, "status", "infrastructureMachineSelector"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noSelector = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineSelector from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	// If one missing but not both, return error
+	if noKind != noSelector {
+		return errors.Errorf("only one of infraMachineKind or infraMachineSelector found, both must be present")
+	}
+
+	// If MachinePool Machines aren't supported by the InfraMachinePool, skip reconcile and return early.
+	if noKind && noSelector {
+		log.V(2).Info("MachinePool Machines not supported, no infraMachineKind or infraMachineSelector found")
+		return nil
+	}
+
+	log.Info("Reconciling MachinePool Machines", "infrastructureMachineKind", infraMachineKind, "infrastructureMachineSelector", infraMachineSelector)
+	var infraMachineList unstructured.UnstructuredList
+
+	infraMachineList.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	infraMachineList.SetKind(infraMachineKind)
+
+	sampleInfraMachine := &unstructured.Unstructured{}
+	sampleInfraMachine.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	sampleInfraMachine.SetKind(infraMachineKind)
+	// Add watcher for infraMachine, if there isn't one already.
+	// Ensure we add a watch to the infraMachine, if there isn't one already.
+	if err := r.externalTracker.Watch(log, sampleInfraMachine, handler.EnqueueRequestsFromMapFunc(infraMachineToMachinePoolMapper)); err != nil {
+		return errors.Wrapf(err, "failed to add watcher on infraMachine %q", sampleInfraMachine.GroupVersionKind())
```

Comment:
```suggestion
		return err
```
nit: I think this is enough. We changed the Watch func a while back to already produce a useful error message, so we don't have to redundantly add the same information here